src/main/java/org/icgc/argo/program_service/services/ProgramService.java

```diff
@@ -138,14 +128,33 @@ public Boolean acceptInvite(UUID invitationId) {
     }
   }
 
-  public void removeProgram(UUID programId) {
-    val program = programRepository.findById(programId);
-    if (program.isPresent()) {
-      removeProgram(program.get());
-    } else {
-      //TODO: add proper error handling
-      log.error("Could not find program {}", programId);
+  public ErrorOr<ProgramEntity> createProgram(@NonNull Program program) {
+    val programEntity = programConverter.programToProgramEntity(program);
+    val now = LocalDateTime.now(ZoneId.of("UTC"));
+    programEntity.setCreatedAt(now);
+    programEntity.setUpdatedAt(now);
+
+    ProgramEntity saved;
+    try {
+      saved = programRepository.save(programEntity);
+    } catch(IOError error) {
+      return new ErrorOr(error);
+    } catch(NoSuchElementException e) {
+      return new ErrorOr(new RuntimeException("Could not save program"));
+    } catch(NullPointerException e) {
+      return new ErrorOr(new Error("Could not save program"));
+    }
+    if (saved == null) {
+      return new ErrorOr(new Error("Could not save program -- no idea why..."));
```

Comment:
there is not clear separation between recoverable and non-recoverable exceptions. Also, the wrapping of exceptions make it a bit difficult to debug because the interpretation is changed. for example, the catch for NosuchElementException, what is that? there is no clear indication at the call site that an Error exception means null pointer exception or how to handle it, and like wise with a runtime exception. There are alot of other exceptions, such as SqlExceptions that are not being caught. I think letting the call site interpret the exception instead of the method interpreting would be the way to go